rollup/internal/controller/relayer/l2_relayer_sanity.go

```diff
@@ -0,0 +1,271 @@
+package relayer
+
+import (
+	"fmt"
+
+	"github.com/scroll-tech/go-ethereum/common"
+	"github.com/scroll-tech/go-ethereum/log"
+
+	"scroll-tech/rollup/internal/orm"
+)
+
+// validateMessageQueueConsistency validates L1 message queue hash consistency
+func (r *Layer2Relayer) validateMessageQueueConsistency(batchIndex uint64, chunks []*orm.Chunk, prevL1MsgQueueHash common.Hash, postL1MsgQueueHash common.Hash) error {
+	if batchIndex == 0 {
+		return nil
+	}
+
+	if len(chunks) == 0 {
+		return fmt.Errorf("batch %d has no chunks for message queue validation", batchIndex)
+	}
+
+	firstChunk := chunks[0]
+	lastChunk := chunks[len(chunks)-1]
+
+	// Calculate total L1 messages in this batch
+	var totalL1MessagesInBatch uint64
+	for _, chunk := range chunks {
+		totalL1MessagesInBatch += chunk.TotalL1MessagesPoppedInChunk
+	}
+
+	// If there were L1 messages processed before this batch, prev hash should not be zero
+	if firstChunk.TotalL1MessagesPoppedBefore > 0 && prevL1MsgQueueHash == (common.Hash{}) {
+		return fmt.Errorf("batch %d prev L1 message queue hash is zero but %d L1 messages were processed before", batchIndex, firstChunk.TotalL1MessagesPoppedBefore)
+	}
+
+	// If there are any L1 messages processed up to this batch, post hash should not be zero
+	totalL1MessagesProcessed := lastChunk.TotalL1MessagesPoppedBefore + lastChunk.TotalL1MessagesPoppedInChunk
+	if totalL1MessagesProcessed > 0 && postL1MsgQueueHash == (common.Hash{}) {
+		return fmt.Errorf("batch %d post L1 message queue hash is zero but %d L1 messages were processed in total", batchIndex, totalL1MessagesProcessed)
+	}
+
+	// Prev and post queue hashes should be different if L1 messages were processed in this batch
+	if totalL1MessagesInBatch > 0 && prevL1MsgQueueHash == postL1MsgQueueHash {
+		return fmt.Errorf("batch %d has same prev and post L1 message queue hashes but processed %d L1 messages in this batch", batchIndex, totalL1MessagesInBatch)
+	}
+
+	return nil
+}
+
+// sanityChecksBeforeConstructingTransaction performs sanity checks before constructing a transaction.
+func (r *Layer2Relayer) sanityChecksBeforeConstructingTransaction(batchesToSubmit []*dbBatchWithChunks) error {
+	if len(batchesToSubmit) == 0 {
+		return fmt.Errorf("no batches to submit")
+	}
+
+	// Basic validation
+	if err := r.validateBatchesBasic(batchesToSubmit); err != nil {
+		return err
+	}
+
+	// Codec version validation
+	if err := r.validateCodecVersions(batchesToSubmit); err != nil {
+		return err
+	}
+
+	// Get previous chunk for continuity check
+	prevChunk, err := r.getPreviousChunkForContinuity(batchesToSubmit[0])
+	if err != nil {
+		return err
+	}
+
+	// Validate each batch in detail
+	if err := r.validateBatchesDetailed(batchesToSubmit, prevChunk); err != nil {
```

Comment:
Why not do the checks of `validateBatchesBasic` and `validateCodecVersions` inside `validateBatchesDetailed` (can rename to `validateBatches`)? I think this would make it a bit easier to read.

Reply:
Sounds good. Combined these small functions in commit: [1b28ff5](https://github.com/scroll-tech/scroll/pull/1714/commits/1b28ff5b5ae43357cd24d27f0eb23a7d4eaefa9a). Previously, this was due to a go lint error. However, making too many small functions seems somewhat overkill.